transaction/DirectedGraph.hpp

```diff
@@ -0,0 +1,173 @@
+/**
+ *   Copyright 2016, Quickstep Research Group, Computer Sciences Department,
+ *     University of Wisconsin—Madison.
+ *
+ *   Licensed under the Apache License, Version 2.0 (the "License");
+ *   you may not use this file except in compliance with the License.
+ *   You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *   Unless required by applicable law or agreed to in writing, software
+ *   distributed under the License is distributed on an "AS IS" BASIS,
+ *   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *   See the License for the specific language governing permissions and
+ *   limitations under the License.
+ **/
+
+#ifndef QUICKSTEP_TRANSACTION_DIRECTED_GRAPH_HPP_
+#define QUICKSTEP_TRANSACTION_DIRECTED_GRAPH_HPP_
+
+#include <algorithm>
+#include <memory>
+#include <unordered_map>
+#include <unordered_set>
+#include <stack>
+#include <vector>
+
+#include "transaction/Transaction.hpp"
+#include "utility/Macros.hpp"
+
+namespace quickstep {
+
+namespace transaction {
+
+/** \addtogroup Transaction
+ *  @{
+ */
+
+/**
+ * @brief Class for representing a directed graph.
+ *        Vertices are transaction ids, edges are
+ *        wait-for relations.
+ */
+class DirectedGraph {
+ public:
+  using NodeId = std::uint64_t;
+
+  /**
+   * @brief Default constructor
+   */
+  DirectedGraph() {}
+
+  /**
+   * @brief Adds node with given transaction id.
+   * @warning It does not check whether transaction id
```

Comment:
To get around this warning, you can implement two methods -- ``addNodeCheckExists`` and ``addNodeUnchecked``. In the first method, you can perform a check before addition of the node. If the check is successful, only then the node will be added. The second method is essentially this method.